File: epi_judge_java/epi/ReverseWords.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TimedExecutor;

public class ReverseWords {

    public static void reverseWords(char[] input) {
        reverse(input, 0, input.length);
        int start = 0;
        while (start < input.length && input[start] == ' ') start++;
        for (int i = 1; i < input.length; i++) {
            if (input[i] != ' ' && input[i - 1] == ' '){
                start = i;
            }
            if ((input[i] == ' ' && input[i - 1] != ' ')) {
                reverse(input, start, i);
                start = -1;
            }
        }
        if (start != -1) reverse(input, start, input.length);
    }

    private static void reverse(char[] arr, int start, int end) {
        int left = start;
        int right = end - 1;
        while (left < right) {
            char temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }

    @EpiTest(testDataFile = "reverse_words.tsv")
    public static String reverseWordsWrapper(TimedExecutor executor, String s)
            throws Exception {
        char[] sCopy = s.toCharArray();

        executor.run(() -> reverseWords(sCopy));

        return String.valueOf(sCopy);
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "ReverseWords.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/reverse_words.cc
#include <string>

#include "test_framework/generic_test.h"
#include "test_framework/timed_executor.h"

using std::string;

void ReverseWords(string &s) {
    std::reverse(s.begin(), s.end());
    int start = 0;
    while (start < s.size() && s[start] == ' ') start++;
    if (start == s.size()) return;
    int end = s.size() - 1;
    while (s[end] == ' ') end--;

    for (int i = start; i <= end; ++i) {
        if (s[i] == ' ' || i == end) {
            std::reverse(s.begin() + start, s.begin() + i + (i == end));
            start = i + 1;
        }
    }
}

string ReverseWordsWrapper(TimedExecutor &executor, string s) {
    string s_copy = s;

    executor.Run([&] { ReverseWords(s_copy); });

    return s_copy;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"executor", "s"};
    return GenericTestMain(args, "reverse_words.cc", "reverse_words.tsv",
                           &ReverseWordsWrapper, DefaultComparator{},
                           param_names);
}
